File: Expendedora.java
import java.util.Scanner;

public class Expendedora {

    
    static int[] codigos = {1, 2, 3, 4};
    static String[] nombres = {"Agua", "Soda", "Papas", "Chocolate"};
    static double[] precios = {10.0, 15.0, 12.5, 18.0};
    static int[] stocks = {5, 5, 5, 5};

    static int[] denominaciones = {10, 5, 1};


    static void mostrarCatalogo() {
        System.out.println("\n--- CATALOGO ---");
        for (int i = 0; i < codigos.length; i++) {
            System.out.printf("[%d] %-10s L. %.2f  (stock: %d)%n",
                    codigos[i], nombres[i], precios[i], stocks[i]);
        }
    }

    static double insertarDinero(double saldoActual, double monto) {
        if (monto <= 0) {
            System.out.println("Monto invalido.");
            return saldoActual;
        }
        return saldoActual + monto;
    }

    static int buscarProducto(int codigo) {
        for (int i = 0; i < codigos.length; i++) {
            if (codigos[i] == codigo) {
                return i;
            }
        }
        return -1;
    }


    static boolean validarSaldo(double precio, double saldo) {
        return saldo >= precio;
    }

    static void calcularVuelto(double monto, int idx) {
        if (monto <= 0 || idx >= denominaciones.length) {
            return; 
        }

        int denom = denominaciones[idx];

        if (monto >= denom) {
            System.out.println("  -> L. " + denom);
            calcularVuelto(monto - denom, idx); 
        } else {
            calcularVuelto(monto, idx + 1); 
        }
    }

    static double seleccionarProducto(int codigo, double saldo) {
        int idx = buscarProducto(codigo);

        if (idx == -1) {
            System.out.println("Producto no encontrado.");
            return saldo;
        }

        if (stocks[idx] <= 0) {
            System.out.println("Producto agotado.");
            return saldo;
        }

        if (!validarSaldo(precios[idx], saldo)) {
            System.out.printf("Saldo insuficiente. Faltan L. %.2f%n", precios[idx] - saldo);
            return saldo;
        }

        stocks[idx]--;
        System.out.println("Entregando: " + nombres[idx]);

        double vuelto = saldo - precios[idx];
        if (vuelto > 0) {
            System.out.printf("Su vuelto es L. %.2f:%n", vuelto);
            calcularVuelto(vuelto, 0);
        }

        return 0.0; 
    }

    static double cancelarTransaccion(double saldo) {
        if (saldo > 0) {
            System.out.printf("Transaccion cancelada. Se devuelven L. %.2f%n", saldo);
        } else {
            System.out.println("No hay saldo que devolver.");
        }
        return 0.0;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        double saldo = 0.0;
        int opcion;

        do {
            System.out.println("\n===== MAQUINA EXPENDEDORA =====");
            System.out.printf("Saldo actual: L. %.2f%n", saldo);
            System.out.println("1. Ver catalogo");
            System.out.println("2. Insertar dinero");
            System.out.println("3. Seleccionar producto");
            System.out.println("4. Cancelar / pedir vuelto");
            System.out.println("0. Salir");
            System.out.print("Opcion: ");
            opcion = sc.nextInt();

            switch (opcion) {
                case 1:
                    mostrarCatalogo();
                    break;
                case 2:
                    System.out.print("Ingrese monto a insertar: ");
                    double monto = sc.nextDouble();
                    saldo = insertarDinero(saldo, monto);
                    break;
                case 3:
                    System.out.print("Ingrese codigo de producto: ");
                    int codigo = sc.nextInt();
                    saldo = seleccionarProducto(codigo, saldo);
                    break;
                case 4:
                    saldo = cancelarTransaccion(saldo);
                    break;
                case 0:
                    System.out.println("Gracias por su compra.");
                    break;
                default:
                    System.out.println("Opcion invalida.");
            }
        } while (opcion != 0);

        sc.close();
    }
}

File: expendedora.py
codigos = [1, 2, 3, 4]
nombres = ["Agua", "Soda", "Papas", "Chocolate"]
precios = [10.0, 15.0, 12.5, 18.0]
stocks = [5, 5, 5, 5]

denominaciones = [10, 5, 1]

def mostrar_catalogo():
    print("\n--- CATALOGO ---")
    for i in range(len(codigos)):
        print(f"[{codigos[i]}] {nombres[i]:<10} L. {precios[i]:.2f}  (stock: {stocks[i]})")


def insertar_dinero(saldo_actual, monto):
    """Recibe el saldo actual y el monto a insertar, devuelve el nuevo saldo."""
    if monto <= 0:
        print("Monto invalido.")
        return saldo_actual
    return saldo_actual + monto


def buscar_producto(codigo):
    """Busca un producto por codigo, devuelve el indice o -1 si no existe."""
    for i in range(len(codigos)):
        if codigos[i] == codigo:
            return i
    return -1


def validar_saldo(precio, saldo):
    """Verifica si el saldo alcanza para el precio dado."""
    return saldo >= precio


def calcular_vuelto(monto, idx=0):
    """Calcula el vuelto de forma RECURSIVA usando las denominaciones disponibles."""
    if monto <= 0 or idx >= len(denominaciones):
        return  # caso base: no queda vuelto o no hay mas denominaciones

    denom = denominaciones[idx]

    if monto >= denom:
        print(f"  -> L. {denom}")
        calcular_vuelto(monto - denom, idx)  # seguimos con la misma denominacion
    else:
        calcular_vuelto(monto, idx + 1)  # pasamos a la siguiente denominacion


def seleccionar_producto(codigo, saldo):
    """Entrega el producto: descuenta stock y muestra el vuelto."""
    idx = buscar_producto(codigo)

    if idx == -1:
        print("Producto no encontrado.")
        return saldo

    if stocks[idx] <= 0:
        print("Producto agotado.")
        return saldo

    if not validar_saldo(precios[idx], saldo):
        print(f"Saldo insuficiente. Faltan L. {precios[idx] - saldo:.2f}")
        return saldo

    stocks[idx] -= 1
    print(f"Entregando: {nombres[idx]}")

    vuelto = saldo - precios[idx]
    if vuelto > 0:
        print(f"Su vuelto es L. {vuelto:.2f}:")
        calcular_vuelto(vuelto)

    return 0.0  


def cancelar_transaccion(saldo):
    """Cancela la transaccion y devuelve el dinero insertado."""
    if saldo > 0:
        print(f"Transaccion cancelada. Se devuelven L. {saldo:.2f}")
    else:
        print("No hay saldo que devolver.")
    return 0.0

def main():
    saldo = 0.0
    opcion = -1

    while opcion != 0:
        print("\n===== MAQUINA EXPENDEDORA =====")
        print(f"Saldo actual: L. {saldo:.2f}")
        print("1. Ver catalogo")
        print("2. Insertar dinero")
        print("3. Seleccionar producto")
        print("4. Cancelar / pedir vuelto")
        print("0. Salir")

        opcion = int(input("Opcion: "))

        if opcion == 1:
            mostrar_catalogo()
        elif opcion == 2:
            monto = float(input("Ingrese monto a insertar: "))
            saldo = insertar_dinero(saldo, monto)
        elif opcion == 3:
            codigo = int(input("Ingrese codigo de producto: "))
            saldo = seleccionar_producto(codigo, saldo)
        elif opcion == 4:
            saldo = cancelar_transaccion(saldo)
        elif opcion == 0:
            print("Gracias por su compra.")
        else:
            print("Opcion invalida.")

if __name__ == "__main__":
    main()